otaclient/app/boot_control/_common.py

```diff
@@ -319,6 +364,223 @@ def reboot(cls):
+
+        # initialize ota_status files if not presented/incompleted/invalid
+        if not (_ota_status and _slot_in_use):
+            logger.info(
+                "ota_status files incompleted/not presented, "
+                "initializing and set status to INITIALIZED..."
+            )
+            _ota_status = wrapper.StatusOta.INITIALIZED
+            self._store_current_slot_in_use(self.active_slot)
+            self._store_current_status(wrapper.StatusOta.INITIALIZED)
+            self._ota_status = _ota_status
+            return
+
+        # updating or rollbacking,
+        # with status==UPDATING or ROLLBACKING, we are expecting this is the first reboot
+        # during switching boot, and we should try finalizing the switch boot.
+        if _ota_status in [wrapper.StatusOta.UPDATING, wrapper.StatusOta.ROLLBACKING]:
+            # finalization failed or not in switching boot
+            if not (
+                self._is_switching_boot(self.active_slot)
+                and self.finalize_switching_boot()
+            ):
+                self._ota_status = (
+                    wrapper.StatusOta.ROLLBACK_FAILURE
+                    if _ota_status == wrapper.StatusOta.ROLLBACKING
+                    else wrapper.StatusOta.FAILURE
+                )
+                self._store_current_status(_ota_status)
+                _err_msg = "finalization failed on first reboot during switching boot"
+                logger.error(_err_msg)
+                raise BootControlInitError(_err_msg)
+
+            # finalizing switch boot successfully
+            self._ota_status = wrapper.StatusOta.SUCCESS
+            self._store_current_status(_ota_status)
+            return
+
+        # status except UPDATING and ROLLBACKING(SUCCESS/FAILURE/ROLLBACK_FAILURE)
+        # remained as it
+        else:
+            self._ota_status = _ota_status
+            self._store_current_status(_ota_status)
+
+            # detect failed reboot, but only print error logging currently
+            if _slot_in_use != self.active_slot:
+                logger.error(
+                    f"boot into old slot {self.active_slot}, "
+                    f"but slot_in_use indicates it should boot into {_slot_in_use}, "
+                    "this might indicate a failed finalization at first reboot after update/rollback"
+                )
+
+    # slot_in_use control
+
+    def _store_current_slot_in_use(self, _slot: str):
+        write_str_to_file_sync(
+            self.current_ota_status_dir / cfg.SLOT_IN_USE_FNAME, _slot
+        )
+
+    def _store_standby_slot_in_use(self, _slot: str):
+        write_str_to_file_sync(
+            self.standby_ota_status_dir / cfg.SLOT_IN_USE_FNAME, _slot
+        )
+
+    def _load_current_slot_in_use(self) -> Optional[str]:
+        if res := read_str_from_file(
+            self.current_ota_status_dir / cfg.SLOT_IN_USE_FNAME, default=""
+        ):
+            return res
```

Comment:
These functions are implemented in SlotInUseMixin, so they can be removed by inheriting it.

Reply:
Emmm I am planning to get rid of all mixin classes though. These methods are core part of ota_status file control logic, so It is better to place them under the same `OTAStatusFilesControl` class.